lib/AST/ASTVerifier.cpp

```diff
@@ -2786,6 +2802,34 @@ class Verifier : public ASTWalker {
         }
       }
 
+      if (FD->getAttrs().hasAttribute<ImplicitlyUnwrappedOptionalAttr>()) {
+        if (!FD->getInterfaceType() ||
+            !FD->getInterfaceType()->is<AnyFunctionType>()) {
+          Out << "Expected FuncDecl to have a function type!\n";
+          abort();
+        }
+
+        auto resultTy =
+            FD->getInterfaceType()->castTo<AnyFunctionType>()->getResult();
+        if (!resultTy) {
+          Out << "Expected result type on function type!\n";
+          abort();
+        }
+
+        // For (Self) -> (A) -> B!, look past the Self param at the
+        // type of the final function call.
+        if (resultTy->is<AnyFunctionType>())
+          resultTy = resultTy->castTo<AnyFunctionType>()->getResult();
```

Comment:
You can call FD->getResultInterfaceType(), which handles unwrapping the extra level of apply for 'self' for you.

(Also this code here is slightly wrong, since you might have a top-level function that returns a function type, in which case you'd be unwrapping it unnecessarily).

Reply:
Great. Yeah I wondered if this was the best way to go about doing this.